syntax = "proto3";
package gladys.v1;

import "common.proto";
import "types.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// ============================================================================
// ORCHESTRATOR SERVICE
// ============================================================================

service OrchestratorService {
    // --------------------------------------------------------------------
    // Event Routing
    // --------------------------------------------------------------------

    // Publish a single event (unary — preferred for most sensors)
    rpc PublishEvent(PublishEventRequest) returns (PublishEventResponse);

    // Publish a batch of events (unary — for high-volume sensors)
    rpc PublishEvents(PublishEventsRequest) returns (PublishEventsResponse);

    // DEPRECATED: Use PublishEvent (single) or PublishEvents (batch) instead.
    // Retained for backward compatibility during migration.
    rpc StreamEvents(stream Event) returns (stream EventAck);

    // Components subscribe to receive events
    rpc SubscribeEvents(SubscribeRequest) returns (stream Event);

    // Subscribe to receive responses
    rpc SubscribeResponses(SubscribeResponsesRequest) returns (stream EventResponse);

    // --------------------------------------------------------------------
    // Component Lifecycle
    // --------------------------------------------------------------------

    // Register a component with the orchestrator
    rpc RegisterComponent(RegisterRequest) returns (RegisterResponse);

    // Unregister (graceful shutdown)
    rpc UnregisterComponent(UnregisterRequest) returns (UnregisterResponse);

    // Send lifecycle command to a component
    rpc SendCommand(CommandRequest) returns (CommandResponse);

    // --------------------------------------------------------------------
    // Health & Status
    // --------------------------------------------------------------------

    // Components send periodic heartbeats
    rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);

    // Get status of all components
    rpc GetSystemStatus(SystemStatusRequest) returns (SystemStatusResponse);

    // Get event queue statistics
    rpc GetQueueStats(GetQueueStatsRequest) returns (GetQueueStatsResponse);

    // List events currently in the queue
    rpc ListQueuedEvents(ListQueuedEventsRequest) returns (ListQueuedEventsResponse);

    // Service health check (unified interface across all services)
    rpc GetHealth(gladys.types.GetHealthRequest) returns (gladys.types.GetHealthResponse);
    rpc GetHealthDetails(gladys.types.GetHealthDetailsRequest) returns (gladys.types.GetHealthDetailsResponse);

    // --------------------------------------------------------------------
    // Service Discovery
    // --------------------------------------------------------------------

    // Resolve component address by ID
    rpc ResolveComponent(ResolveRequest) returns (ResolveResponse);
}

// ----------------------------------------------------------------------------
// Event Routing Messages
// ----------------------------------------------------------------------------

message EventAck {
    string event_id = 1;
    bool accepted = 2;
    string error_message = 3;
    string response_id = 4;
    string response_text = 5;
    float predicted_success = 6;
    float prediction_confidence = 7;
    bool routed_to_llm = 8;
    string matched_heuristic_id = 9;
    bool queued = 10;  // True if event was queued for async processing
}

message PublishEventRequest {
    Event event = 1;
    RequestMetadata metadata = 15;
}

message PublishEventResponse {
    EventAck ack = 1;
}

message PublishEventsRequest {
    repeated Event events = 1;
    RequestMetadata metadata = 15;
}

message PublishEventsResponse {
    int32 accepted_count = 1;
    repeated EventError errors = 2;
}

message EventError {
    string event_id = 1;
    string error_message = 2;
}

message SubscribeRequest {
    string subscriber_id = 1;
    repeated string source_filters = 2;
    repeated string event_types = 3;
}

message SubscribeResponsesRequest {
    string subscriber_id = 1;
    repeated string source_filters = 2;
    bool include_immediate = 3;
}

message EventResponse {
    string event_id = 1;
    string response_id = 2;
    string response_text = 3;
    float predicted_success = 4;
    float prediction_confidence = 5;
    RoutingPath routing_path = 6;
    string matched_heuristic_id = 7;
    int64 event_timestamp_ms = 8;
    int64 response_timestamp_ms = 9;
    string source = 10;           // Sensor ID from original event
    string original_text = 11;    // Event raw_text from original event
}

enum RoutingPath {
    ROUTING_PATH_UNSPECIFIED = 0;
    ROUTING_PATH_IMMEDIATE = 1;
    ROUTING_PATH_QUEUED = 2;  // Renamed from ACCUMULATED - events processed via queue
}

// ----------------------------------------------------------------------------
// Registration Messages
// ----------------------------------------------------------------------------

message RegisterRequest {
    string component_id = 1;
    string component_type = 2;
    string address = 3;
    ComponentCapabilities capabilities = 4;
    RequestMetadata metadata = 15;
}

message ComponentCapabilities {
    TransportMode transport_mode = 1;
    int32 batch_size = 2;
    int32 batch_interval_ms = 3;
    bool configurable = 4;
    repeated string supported_instructions = 5;
    InstancePolicy instance_policy = 6;
}

enum TransportMode {
    TRANSPORT_MODE_UNSPECIFIED = 0;
    TRANSPORT_MODE_STREAMING = 1;
    TRANSPORT_MODE_BATCHED = 2;
    TRANSPORT_MODE_EVENT = 3;
}

enum InstancePolicy {
    INSTANCE_POLICY_SINGLE = 0;
    INSTANCE_POLICY_MULTIPLE = 1;
}

message RegisterResponse {
    bool success = 1;
    string error_message = 2;
    string assigned_id = 3;
}

message UnregisterRequest {
    string component_id = 1;
    RequestMetadata metadata = 15;
}

message UnregisterResponse {
    bool success = 1;
}

// ----------------------------------------------------------------------------
// Command Messages
// ----------------------------------------------------------------------------

message CommandRequest {
    string target_component_id = 1;
    Command command = 2;
    google.protobuf.Struct args = 3;
    RequestMetadata metadata = 15;
}

enum Command {
    COMMAND_UNSPECIFIED = 0;
    COMMAND_START = 1;
    COMMAND_STOP = 2;
    COMMAND_PAUSE = 3;
    COMMAND_RESUME = 4;
    COMMAND_RELOAD = 5;
    COMMAND_HEALTH_CHECK = 6;
    COMMAND_RECOVER = 7;
}

message CommandResponse {
    bool success = 1;
    string error_message = 2;
    ComponentStatus status = 3;
}

// ----------------------------------------------------------------------------
// Health Messages
// ----------------------------------------------------------------------------

message HeartbeatRequest {
    string component_id = 1;
    ComponentState state = 2;
    string error_message = 3;
    RequestMetadata metadata = 15;
}

message HeartbeatResponse {
    bool acknowledged = 1;
    repeated PendingCommand pending_commands = 2;
}

message PendingCommand {
    string command_id = 1;
    Command command = 2;
    google.protobuf.Struct args = 3;
}

message SystemStatusRequest {
    RequestMetadata metadata = 15;
}

message SystemStatusResponse {
    repeated ComponentStatus components = 1;
    google.protobuf.Timestamp timestamp = 2;
}

message GetQueueStatsRequest {}

message GetQueueStatsResponse {
    int32 queue_size = 1;              // Current number of pending events
    int64 total_queued = 2;            // Lifetime total events queued
    int64 total_processed = 3;         // Lifetime total events processed
    int64 total_timed_out = 4;         // Lifetime total events timed out
    float avg_wait_time_ms = 5;        // Average time events wait in queue
    float processing_rate_per_sec = 6; // Events/second throughput
}

message ListQueuedEventsRequest {
    int32 limit = 1;              // Max events to return (0 = all)
}

message ListQueuedEventsResponse {
    repeated QueuedEventInfo events = 1;
    int32 total_count = 2;        // Total events in queue (may exceed limit)
}

message QueuedEventInfo {
    string event_id = 1;
    string source = 2;
    string event_type = 3;
    float salience = 4;
    int64 enqueue_time_ms = 5;
    int64 age_ms = 6;             // How long in queue
    string matched_heuristic_id = 7;
    float heuristic_confidence = 8;
    string raw_text = 9;          // Event text for display
}

// ----------------------------------------------------------------------------
// Service Discovery Messages
// ----------------------------------------------------------------------------

message ResolveRequest {
    string component_id = 1;
    string component_type = 2;
    RequestMetadata metadata = 15;
}

message ResolveResponse {
    bool found = 1;
    string address = 2;
    ComponentCapabilities capabilities = 3;
}
